torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -232,21 +234,26 @@ def powerSGD_hook(state: PowerSGDState, bucket) -> torch.futures.Future:
         p_idx += n * matrix_approximation_rank
         q_idx += m * matrix_approximation_rank
 
-    # Initialize and then orthogonalize Qs.
-    with torch.random.fork_rng(devices=[]):
-        # Fork this RNG to avoid changing the seed globally and affecting the random sampling anywhere else in the training.
-        # The seed makes sure that the initial random values are the same across all the DDP replicas.
-        # Such seed should differ at every step.
-        # Since it is very slow to fork RNG state across all the CUDA devices,
-        # only fork on CPU and then move the generated tensor to the CUDA device.
-        torch.manual_seed(state.rng.randint(1_000_000_000))
+    # If warm-start is enabled, reuse Qs from the previous iteration if possible and skip filling random values.
+    # The exceptions are the first time and when the buckets are rebuilt.
+    if not need_randomize_qs:
         for q in qs:
-            q.data = torch.randn(
-                *q.shape,
-                device="cpu",
-                dtype=dtype,
-            ).to(device)
             _orthogonalize(q)
+    else:
+        with torch.random.fork_rng(devices=[]):
+            # Fork this RNG to avoid changing the seed globally and affecting the random sampling anywhere else in the training.
+            # The seed makes sure that the initial random values are the same across all the DDP replicas.
+            # Such seed should differ at every step.
+            # Since it is very slow to fork RNG state across all the CUDA devices,
+            # only fork on CPU and then move the generated tensor to the CUDA device.
+            torch.manual_seed(state.rng.randint(1_000_000_000))
+            for q in qs:
+                q.data = torch.randn(
+                    *q.shape,
+                    device="cpu",
+                    dtype=dtype,
+                ).to(device)
+                _orthogonalize(q)
```

Comment:
Nit: we call `_orthogonalize(q)` in both branches, will it be clearer to do:
```
for q in qs:
    _orthog(q) 
```

only once after the if/else block?

Reply:
> Nit: we call `_orthogonalize(q)` in both branches, will it be clearer to do:
> 
> ```
> for q in qs:
>     _orthog(q) 
> ```
> 
> only once after the if/else block?

Agree that calling `_orthogonalize(q)` in a loop in just one branch can make the code more clear. However, this means two loops if warm-start is disabled: 1) create matrices; 2) orthogonalize these matrices.


In contrast, the current implementation can leverage the data locality, and hence it's better for the performance -- once q is created with random values, orthogonalize q immediately when it's still hot (i.e., in cache).